astroquery/vo_service_discovery/core.py

```diff
@@ -0,0 +1,250 @@
+            raise ValueError("""
+            service_type must be one of conesearch, simpleimageaccess, simplespectralaccess, tableaccess, or
+            their alternate strings: cone, image, spectra, spectrum, table or tap.
+            """)
+
+        query_retcols = """
+          select res.waveband,res.short_name,cap.ivoid,res.res_description,
+          intf.access_url,res.reference_url,res_role.role_name as publisher,cap.cap_type as service_type
+          from rr.capability as cap
+            natural join rr.resource as res
+            natural join rr.interface as intf
+            natural join rr.res_role as res_role
+            """
+
+        query_where = " where "
+
+        wheres = []
+        if service_type != "":
+            wheres.append("cap.cap_type like '%{}%'".format(service_type))
+
+        # Currently not supporting SIAv2 in SIA library.
+        if service_type == 'simpleimageaccess':
+            wheres.append("standard_id != 'ivo://ivoa.net/std/sia#query-2.0'")
+        if source != "":
+            wheres.append("cap.ivoid like '%{}%'".format(source))
+        if waveband != "":
+            if ',' in waveband:
+                allwavebands = []
+                for w in waveband.split(','):
+                    allwavebands.append("res.waveband like '%{}%' ".format(w).strip())
+                wheres.append("(" + " or ".join(allwavebands) + ")")
+            else:
+                wheres.append("res.waveband like '%{}%'".format(waveband))
+
+        wheres.append("res_role.base_role = 'publisher'")
+        if publisher != "":
+            wheres.append("res_role.role_name like '%{}%'".format(publisher))
+
+        if keyword != "":
+            keyword_where = """
+             (res.res_description like '%{}%' or
+            res.res_title like '%{}%' or
+            cap.ivoid like '%{}%')
+            """.format(keyword, keyword, keyword)
+            wheres.append(keyword_where)
+
+        query_where = query_where+logic_string.join(wheres)
+
+        if order_by != "":
+            query_order = "order by {}".format(order_by)
+        else:
+            query_order = ""
+
+        query = query_retcols+query_where+query_order
+
+        return query
+
+    def _build_counts_adql(self, field, minimum=1):
+
+        field_table = None
+        field_alias = field
+        query_where_filter = ''
+        if field.lower() == 'waveband':
+            field_table = 'rr.resource'
+        elif field.lower() == 'publisher':
+            field_table = 'rr.res_role'
+            field = 'role_name'
+            query_where_filter = ' where base_role = \'publisher\' '
+        elif field.lower() == 'service_type':
+            field_table = 'rr.capability'
+            field = 'cap_type'
+
+        if field_table is None:
+            return None
+        else:
+            query_select = 'select ' + field + ' as ' + field_alias + ', count(' + field + ') as count_' + field_alias
```

Comment:
this may be easier to read if `.format()` is used rather than string additions

Reply:
I reworked this section using format(), and I agree it looks a bit better.